Improved Effective Moisture Penetration Depth model

nmerket opened the pull request:
[New Feature Proposal](https://github.com/NREL/EnergyPlus/blob/empd/design/FY2015/EMPD2.md)

Todo:
- [x] Describe methodology in NFP/Design Doc
- [x] Code the feature
  - [x] Fix initialization bug
- [x] Documentation
  - [x] Last updates to docs by @jdwoods21 
- [x] Unit Tests
  - [x] Fix Room Airflow Network unit test
- [x] Integration/Regression Test Files
  - [x] Update with more reasonable inputs for new model
- [x] Transition rules/equations
  - [x] Transition program (Edwin)
- [x] Assess and modify outputs
- [x] Review

@jdwoods21 @nealkruis @Myoldmopar @rongpengzhang @lgu1234

Commit:
adding NFP

Myoldmopar commented:
Nice!!

Myoldmopar commented:
@nmerket Can you pull develop into this branch once again?  I just added the FY2015 folder to develop because I realized if multiple people did that we could end up with tree conflicts.

Commit:
Merge branch 'develop' into empd

Commit:
porting @jdwoods21 empd code from 7.1

nmerket commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -5086,6 +5092,8 @@ CalcHeatBalanceInsideSurf( Optional_int_const ZoneToResimulate ) // if passed in
 			Real64 const HConvIn_surf( HConvInFD( SurfNum ) = HConvIn( SurfNum ) );
 			RhoVaporAirIn( SurfNum ) = min( PsyRhovFnTdbWPb_fast( MAT_zone, ZoneAirHumRat_zone, OutBaroPress ), PsyRhovFnTdbRh( MAT_zone, 1.0, HBSurfManInsideSurf ) );
+			// TODO: Check with Jason about why RhoVaporAirIn( SurfNum ) wasn't added in his version.
+			//       The original was commented out, which makes me think it was a debugging effort.
```
@jdwoods21 Your version had the `RhoVaporAirIn( SurfNum )` removed, should this be deleted here too?

nmerket commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -5646,7 +5654,7 @@ CalcHeatBalanceInsideSurf( Optional_int_const ZoneToResimulate ) // if passed in
 				RhoVaporSurfIn( SurfNum ) = MoistEMPDNew( SurfNum );
 				//SUMC(ZoneNum) = SUMC(ZoneNum)-MoistEMPDFlux(SurfNum)*Surface(SurfNum)%Area
-
+				// TODO: Check with Jason that these equations were applied correctly. It's different than in his code.
```
@jdwoods21 Your version had these equations as 

``` FORTRAN
  !       SumHmAW(ZoneNum)= SumHmAW(ZoneNum)+HM(SurfNum)*Surface(SurfNum)%Area*  &
  !                         (Wwall(SurfNum)-ZoneAirHumRat(ZoneNum))
  !       SumHmARa(ZoneNum)= SumHmARa(ZoneNum)+HM(SurfNum)*Surface(SurfNum)%Area
  !       SumHmARaW(ZoneNum)= SumHmARaW(ZoneNum)+HM(SurfNum)*Surface(SurfNum)%Area*Wwall(SurfNum)
            SumHmAW(ZoneNum)= SumHmAW(ZoneNum) + FluxZone(SurfNum)*Surface(SurfNum)%Area !SumHmAW(ZoneNum)+HMassConvInFD(SurfNum)*Surface(SurfNum)%Area*  &
                    !       (RhoVaporSurfIn(SurfNum)-RhoVaporAirIn(SurfNum))

        RhoAirZone=PsyRhoAirFnPbTdbW(OutBaroPress,Mat(Surface(SurfNum)%Zone),PsyWFnTdbRhPb(Mat(Surface(SurfNum)%Zone),  &
              PsyRhFnTdbRhov(Mat(Surface(SurfNum)%Zone),RhoVaporAirIn(SurfNum),'RhoAirZone'),OutBaroPress))

         Wsurf=PsyWFnTdbRhPb(TempSurfInTmp(SurfNum), &
              PsyRhFnTdbRhov(TempSurfInTmp(SurfNum),RhoVaporSurfIn(SurfNum),'Wsurf'),OutBaroPress)

          SumHmARa(ZoneNum)= SumHmARa(ZoneNum)+HMshort(SurfNum)*Surface(SurfNum)%Area*RhoAirZone

         SumHmARaW(ZoneNum)= SumHmARaW(ZoneNum)+HMshort(SurfNum)*Surface(SurfNum)%Area*RVsurface(SurfNum)
```

Can you check and make sure they are applied correctly here?

jdwoods21 committed:
Update to correct EMPD2 equations

Commit:
clean up and reference use in CalcMoistureBalanceEMPD

jdwoods21 committed:
Latest updates to EMPD eq'ns

Commit:
Updating empd input processing. Calc methods still using old ones. Won't compile.

jdwoods21 committed:
EMPD with new inputs

Updated equations to accept Noel's input processing. Working as
expected.

jdwoods21 committed:
Temporary commit

jdwoods21 committed:
Fixes and clean-up of EMPD eq'ns.

- Removed temporary variables used for debugging
- Reverted back to equations including latent-sensible heat conversion
- Fixed mass transfer equation in HAMT eq'ns
- Added comments to MoistureBalanceEMPDManager.cc
- Still need to add EMPD output variables

nmerket commented on src/EnergyPlus/DataMoistureBalanceEMPD.hh:
```diff
@@ -32,6 +32,15 @@ namespace DataMoistureBalanceEMPD {
 	extern Array1D< Real64 > MoistEMPDNew; // Moisture level at interior surfaces at current interation
```
@jdwoods21 Which of these do we not need anymore? From what I can see, I should be able to get rid of:

- `MoistEMPDOld`
- `MoistEMPDInt`
- `MoistEMPDNew`
- `MoistEMPDFlux`

Is that right?

Commit:
Merge branch 'develop' into empd

Commit:
removing unused variables

Commit:
removing another unused array

Commit:
removing unused function

Commit:
unit test

Commit:
merging in develop

Commit:
renaming some variables

Commit:
updating reporting variables

Commit:
clean up of unnecessary gloabl arrays

Commit:
fixing bad header include

Commit:
updating unit test to include deep layer calcs

Commit:
Merge branch 'develop' into empd

Commit:
transitioning the old to new empd model in the examples

Commit:
Merge remote-tracking branch 'origin/develop' into empd

nmerket commented on src/Transition/transition_empd.py:
```diff
@@ -0,0 +1,177 @@
+# This is a quick and dirty python script to transition the EMPD model.
+# You'll want to use this math in the official transition program.
+# Once you're done that, feel free to delete this.
```
@Myoldmopar whenever you get around to porting this to the new Fortran transition program, here's what you'll need to know to transition the inputs for the EMPD model.

Myoldmopar replied:
!!!!!! :snake:

Commit:
updating docs

Commit:
Merge remote-tracking branch 'origin/develop' into empd

nmerket commented on tst/EnergyPlus/unit/RoomAirflowNetwork.unit.cc:
```diff
@@ -249,8 +249,8 @@ TEST_F( RoomAirflowNetworkTest, RAFNTest )
 	Surface( 1 ).HeatTransferAlgorithm = HeatTransferModel_EMPD;
 	Surface( 2 ).HeatTransferAlgorithm = HeatTransferModel_EMPD;
-	MoistEMPDNew( 1 ) = 0.0011;
-	MoistEMPDNew( 2 ) = 0.0012;
```
@lgu1234 with the rearranging I've been doing in the EMPD model, the `RoomAirflowNetwork` unit test is failing now because it touches EMPD. I'm going to try to take a look and figure out what's going on, but any help you could lend on this would be really appreciated.

lgu1234 replied:
@nmerket Here is a section of code in the CalcSurfaceMoistureSums function of the RoomAirflowNetwork module

```
        if ( Surface( SurfNum ).HeatTransferAlgorithm == HeatTransferModel_EMPD ) {

            UpdateMoistureBalanceEMPD( SurfNum );
            RhoVaporSurfIn( SurfNum ) = MoistEMPDNew( SurfNum );

            SumHmAW = SumHmAW + HMassConvInFD( SurfNum )*Surface( SurfNum ).Area* ( RhoVaporSurfIn( SurfNum ) - RhoVaporAirIn( SurfNum ) );
            SumHmARa = SumHmARa + HMassConvInFD( SurfNum )*Surface( SurfNum ).Area* PsyRhoAirFnPbTdbW( OutBaroPress, TempSurfInTmp( SurfNum ),
                PsyWFnTdbRhPb( TempSurfInTmp( SurfNum ), PsyRhFnTdbRhovLBnd0C( TempSurfInTmp( SurfNum ), RhoVaporAirIn( SurfNum ) ), OutBaroPress ) );
            SumHmARaW = SumHmARaW + HMassConvInFD( SurfNum )*Surface( SurfNum ).Area*RhoVaporSurfIn( SurfNum );
        }
```

nmerket replied:
@lgu1234 Thanks, figured it out.

lgu1234 replied:
@nmerket Great!

Commit:
Merge remote-tracking branch 'origin/develop' into empd

Commit:
fixing room airflow network unit test

nmerket commented:
@jdwoods21  `EMPD5ZoneWaterCooled_HighRHControl.idf` was the test file causing problems.

Commit:
Merge remote-tracking branch 'origin/develop' into empd

Commit:
fixing import issues

Commit:
fixing initialization error

Commit:
fixing empd init and unit test

Commit:
updating the moisture materials dataset file

nmerket commented on datasets/MoistureMaterials.idf:
```diff
@@ -6,85 +6,125 @@
+    0,                        !- Deep-layer penetration depth {m}
+    0,                        !- Coating layer thickness {m}
+    0;                        !- Coating layer water vapor diffusion resistance factor {m}
+    
```
@jdwoods21, I just found this file that needed to be updated. It's supposed to contain EMPD input data for a collection of materials. I went ahead and updated it using our transition script and some material density data from `datasets/ASHRAE_2005_HOF_Materials.idf`, Wolfram Alpha (vinyl), and Wikipedia (soil). I don't know if you need to update these with appropriate deep layer stuff.

Commit:
updating to full design doc

Commit:
Merge remote-tracking branch 'origin/develop' into empd

nrel-bot-3 commented:
@nmerket @lgentile it has been 14 days since this pull request was last updated.

Commit:
Merge branch 'develop' into empd

Commit:
updating test and dataset files to have deep layers where appropriate

Commit:
Merge remote-tracking branch 'origin/develop' into empd

Myoldmopar commented:
@nmerket OK, I'm reviewing this now.  First step, trying to bring up to develop so that I can work on the transition program with the latest changes.  Only one conflict, that's good.  Here it is:

```
<<<<<<< HEAD
    ~RoomAirflowNetworkTest( ) // Reset global state
    {
        NumOfZones = 0;
        NumOfNodes = 0;
        BeginEnvrnFlag = false;
        RoomAirflowNetworkZoneInfo.clear( );
        ZoneEquipConfig.clear( );
        ZoneEquipList.clear( );
        Zone.clear( );
        ZoneIntGain.clear( );
        NodeID.clear( );
        Node.clear( );
        Surface.clear( );
        HConvIn.clear( );
        TempSurfInTmp.clear( );
        RVSurface.clear( );
        RVSurfaceOld.clear( );
        RVDeepLayer.clear( );
        RVdeepOld.clear( );
        RVSurfLayerOld.clear( );
        RVSurfLayer.clear( );
        RhoVaporSurfIn.clear( );
        RhoVaporAirIn.clear( );
        HMassConvInFD.clear( );
        MAT.clear( );
        ZoneAirHumRat.clear( );
        AirflowNetworkLinkageData.clear( );
        AirflowNetworkNodeSimu.clear( );
        AirflowNetworkLinkSimu.clear( );
        RAFN.clear( );
=======
    virtual void TearDown() {
        EnergyPlusFixture::TearDown();  // Remember to tear down the base fixture after cleaning up derived fixture!
>>>>>>> origin/develop
```

OK.  So, it's a little unclear what happened.  As far as I can tell, in this branch, you had added some additional [setup](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-8ac2ff96faa207dcff2c8ab37958783eR76) and [teardown](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-8ac2ff96faa207dcff2c8ab37958783eR109).  But then somewhere, @mbadams5 went into this unit test, and appears to have [pulled all the teardown out](https://github.com/NREL/EnergyPlus/commit/61d212f6684baad00a1ef80be93886be31403f44#diff-8ac2ff96faa207dcff2c8ab37958783eR88), and added it most likely to the `clear_state` in that particular namespace.  I think this is probably just a matter of accepting @mbadams5 change here in this file, but making sure the additional new teardown functionality is included in the main source file `clear_state` function.  Would you mind doing that fairly simple merge before I continue to any other aspect here?  After that I'll make up the transition program based on your script and we can continue the review process.

mbadams5 commented:
@Myoldmopar Yes I removed the custom destructor in favor the calling the `EnergyPlusFixture::TearDown();` since that makes sure ALL global data including what was included in the destructor is reset.

nmerket commented:
Thanks @mbadams5, I'll make that merge then and post it back here.

mbadams5 commented:
@nmerket Just make sure to add those new variables to `clear_state()`

Commit:
Merge remote-tracking branch 'origin/develop' into empd

nmerket commented:
Oh, okay I thought you did that. Hang on.

mbadams5 commented:
I did it with the old variables but not the new ones you added in this PR.

Commit:
adding clear_state to MoistureBalanceEMPDManager

mbadams5 committed:
Fix clear_state and minor cleanups

nrel-bot-3 commented:
@nmerket @lgentile it has been 14 days since this pull request was last updated.

Commit:
Merge remote-tracking branch 'origin/develop' into empd

nmerket commented:
@Myoldmopar, just wanted to make sure this is still on your radar.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into empd

Myoldmopar committed:
Update Transition source for new empd model

Myoldmopar committed:
Add transition rules into workbook

Myoldmopar commented:
OK.  Added transition such and such.  @nmerket Please check out the src/Transition/InputRulesFiles/Rules...xls file, including the second sheet in the workbook, and make sure everything looks good to go.

I also pulled the latest develop in, so if CI comes back clean...maybe this goes right in.

Myoldmopar commented on src/Transition/UtilityRoutines.f90:
```diff
@@ -895,6 +895,73 @@ SUBROUTINE ShowErrorMessage(ErrorMessage,OutUnit1,OutUnit2)
+  TK = Tdb + 273.15      ! Converts from degC to degK
+
+  IF (TK <= 273.15) THEN
+         retval = EXP(C1/TK + C2 + C3*TK + C4*TK**2 + C5*TK**3 + C6*TK**4 + C7*LOG(TK)) / 1000
```
Just to **ensure** no shenanigans, I verified the nature of logarithms in each language:
- Python math.log(x[, base]): [With one argument, return the natural logarithm of x (to base e).](https://docs.python.org/2/library/math.html#math.log)
- Fortran LOG(x): [natural logarithm of x](http://www.cs.mtu.edu/~shene/COURSES/cs201/NOTES/chap02/funct.html)

OK, we're good, carry on.

Myoldmopar commented:
Final review points now that the transition work is complete:
- [x] Could you add some \note entry to [this field](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-23ccf090b80d26e885712256b9a6d888R6835)?  I know the docs will have full details, but if a few words might help describe it, that would be nice for interfaces that mine the idd for info
- [x] The field names [here](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-23ccf090b80d26e885712256b9a6d888R6856) should be Title Case, and @mjwitte may have a comment regarding the hyphen in `surface-layer` and `deep-layer`.
- [x] Any reason you removed [this line](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-c10afc2ba60f4d2552b6510cdbbf5a95L4940)?
- [x] Padding for parentheses got nuked, probably by Visual Studio, [here](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-c10afc2ba60f4d2552b6510cdbbf5a95R5256)
- [x] Indent lost [here](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-691a3b35c3d65b682c8cf7acda7ffac9R557)
- [x] Were [these](https://github.com/NREL/EnergyPlus/pull/5093/files#diff-ff08ad5fe0fccbde744effe7442491b1L9159) not being used?  Or is the initialization back to zero done somewhere else?

I don't have any significant comments otherwise.  If you can address these, and anything @mjwitte adds to the review, this can merge anytime.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -6850,10 +6849,31 @@ MaterialProperty:MoisturePenetrationDepth:Settings,
        \required-field
        \units dimensionless
        \type real
+  N6 , \field Surface-layer penetration depth
```
@Myoldmopar Hyphens are used sparingly in field names, most common being Dry-Bulb.  My preference here would be to drop the hyphens, but if you keep it, both words should be Title case, "Surface-Layer Penetration Depth".
And be sure to update the I/O ref headings to match.
IP-Units for m will default to ft. If you prefer inches, then add "\ip-units in" to these fields.

jdwoods21 replied:
@nmerket 

Re: First comment from @Myoldmopar, \note should be:
\Ratio of water vapor permeability of stagnant air to water vapor permeability of material

mjwitte commented on doc/src/docs/InputOutputReference/01a-InputOutputReference.md:
```diff
@@ -4412,119 +4412,179 @@ Material:AirGap,B1 - AIRSPACE RESISTANCE,  ! Material Name
 ### Moisture Penetration Depth (EMPD) Outputs
 
-Output variables applicable to heat transfer surfaces using EMPD model:
+#### EMPD surface inside face water vapor density [kg/m^3]
```
No superscript ^ symbols here and below in the headings, simply [kg/m3] so it matches the source and actual output.

mjwitte commented on doc/src/docs/InputOutputReference/01a-InputOutputReference.md:
```diff
@@ -4412,119 +4412,179 @@ Material:AirGap,B1 - AIRSPACE RESISTANCE,  ! Material Name
-    0.9200000;               !- Visible Absorptance
+$\frac{du}{d\phi}$ = slope of moisture soprtion curve, $a b  \phi^{b-1} + c d \phi^{d-1}$
 
+#### Field: Deep-layer penetration depth
```
I don't see the units stated for this.  Check other fields as well.

mjwitte commented on src/EnergyPlus/MoistureBalanceEMPDManager.cc:
```diff
@@ -340,48 +358,52 @@ namespace MoistureBalanceEMPDManager {
+			if ( Surface( SurfNum ).Class == SurfaceClass_Window ) continue;
+			EMPDReportVarsData & rvd = EMPDReportVars( SurfNum );
+			const std::string surf_name = Surface( SurfNum ).Name;
+			SetupOutputVariable( "EMPD Surface Inside Face Water Vapor Density [kg/m^3]", rvd.rv_surface, "Zone", "State", surf_name );
```
No carats in the units here and below.

mjwitte commented on testfiles/1ZoneUncontrolledFourAlgorithms.idf:
```diff
@@ -502,13 +502,17 @@
+    0.397173,                   !- Moisture Equation Coefficient b {dimensionless}
+    0.007774,                   !- Moisture Equation Coefficient c {dimensionless}
+    11.7057,                   !- Moisture Equation Coefficient d {dimensionless}
+    0.004,                   !- Surface-layer penetration depth {m}
```
Be sure to update field names in test files to match changes in the IDD, just for consistency.

nmerket commented on src/EnergyPlus/SurfaceGeometry.cc:
```diff
@@ -9156,15 +9156,6 @@ namespace SurfaceGeometry {
-						Material( TotMaterials ).EMPDaCoeff = 0.0;
-						Material( TotMaterials ).EMPDbCoeff = 0.0;
-						Material( TotMaterials ).EMPDcCoeff = 0.0;
-						Material( TotMaterials ).EMPDdCoeff = 0.0;
```
@Myoldmopar These are initialized to zero [here](https://github.com/NREL/EnergyPlus/blob/4db1c6507021456b6879198bbdb475496a6db789/src/EnergyPlus/DataHeatBalance.hh#L875-L883).

Commit:
Clean up edits based on reviewer comments

@Myoldmopar
@mjwitte

Commit:
Merge remote-tracking branch 'origin/develop' into empd

nmerket commented:
@Myoldmopar @mjwitte I believe I have addressed all your comments and this is ready to merge.

mjwitte commented on datasets/MoistureMaterials.idf:
```diff
@@ -6,84 +6,100 @@
+    0.939,                    !- Moisture Equation Coefficient b {dimensionless}
+    0.0202,                   !- Moisture Equation Coefficient c {dimensionless}
+    12.2,                     !- Moisture Equation Coefficient d {dimensionless}
+    0.019,                    !- Surface-layer penetration depth {m}
```
@nmerket Thanks for the IDD cleanup, forgot to mention the dataset field names.

mjwitte commented on doc/src/docs/InputOutputReference/01a-InputOutputReference.md:
````diff
@@ -4412,119 +4412,179 @@ Material:AirGap,B1 - AIRSPACE RESISTANCE,  ! Material Name
+This set of inputs can be used for aerated concrete (assuming linear sorption curve). Density, input elsewhere, is :
+
+```idf
+MaterialProperty:MoisturePenetrationDepth:Settings,
````
And the idf snippet field names in the I/O Ref.  And one other nit-pick, the output variable names below should be title case (the SetupOutput calls have it correct).

mjwitte commented:
Also, output variable names have changed.  These need to be documented in [Report Variables 8-4-0 to 8-5-0.csv](https://github.com/NREL/EnergyPlus/blob/empd/src/Transition/SupportFiles/Report%20Variables%208-4-0%20to%208-5-0.csv)  If there's a new one that produces equivalent values, then we can map oldname to newname and it should be fine.  But if there isn't a new equivalent, then we should probably tag it as delete, so the missing output gets the user's attention?  @Myoldmopar ?

Commit:
updating input names in docs and datasets

nmerket commented:
@mjwitte When editing [this file](https://github.com/NREL/EnergyPlus/blob/empd/src/Transition/SupportFiles/Report%20Variables%208-4-0%20to%208-5-0.csv) should I add lines for new report variables, or just ones that have changed names or have been deleted?

mjwitte commented:
Just changes or deletions.  This gets used by transition to change or delete existing output:variable objects in old idfs.

Commit:
correct output variable names in docs

Commit:
adding removed output variables to transition file

nmerket commented:
Alright, that should do it.

mjwitte commented:
@nmerket Thanks.  @Myoldmopar All of my comments have been addressed, so this is back to you.  Question - should the py script be deleted now that the transition routine has been written?  I did not test the transition . . .

Myoldmopar commented:
I tested the transition, it should be good to go.  I think I'll delete the Python script, it is included as a worksheet in the main transition workbook.

Myoldmopar committed:
Delete transition_empd.py

Myoldmopar commented:
As long as CI comes back happy this evening, I'll merge this in.

nmerket commented:
@Myoldmopar, hold tight. We've found a bug in the new EMPD code that I'd like to fix before merging. I'll let you know when it's ready again.

Myoldmopar commented:
k

Myoldmopar commented:
@nmerket Do you anticipate having the bug cleaned up today?  We'll be frozen at some point today.

nmerket commented:
I really hope so. Are you around today? I might want some help figuring out what's going on.

Myoldmopar commented:
I'm around this morning, come over now if you want.

Commit:
using values from previous timestep

also adding some print statements for debugging

Commit:
Merge branch 'empd' of github.com:NREL/EnergyPlus into empd

Myoldmopar commented:
@nmerket Since we are after the freeze, this isn't urgent anymore, but I'm still anxious to get it wrapped up.  Did that last commit you made address the problem you were seeing or does this still need work?

nmerket commented:
No, it still needs some work. @jdwoods21 and I are meeting next week to figure it out.

nrel-bot-3 commented:
@Myoldmopar @nmerket @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@Myoldmopar @nmerket @lgentile it has been 14 days since this pull request was last updated.

nmerket commented:
@Myoldmopar can you port my docs changes over to latex when you get a chance?

Commit:
Merge commit '79b6d8e7119012f2396bbcdc9c5f801d290d6e8c' into empd

nmerket commented:
I merged up the [commit before the latex docs merge](https://github.com/NREL/EnergyPlus/commit/79b6d8e7119012f2396bbcdc9c5f801d290d6e8c) on develop to hopefully make updating the docs easier.

nmerket commented on src/Transition/CreateNewIDFUsingRulesV8_5_0.f90:
```diff
@@ -363,6 +365,49 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
                 nodiff=.false.
 
+    !!!    Changes for this version
+              CASE('MATERIALPROPERTY:MOISTUREPENETRATIONDEPTH:SETTINGS')
```
It looks like this merged okay, but it should probably be in a new `CreateNewIDFUsingRulesV8_6_0.f90` source file instead of here. In an effort not to screw things up, I'm going to leave that you @Myoldmopar.

Commit:
Merge branch 'develop' into empd

Migrate doc changes from markdown to latex, phew.

Commit:
Merge remote-tracking branch 'origin/develop' into empd

nrel-bot-3 commented:
@Myoldmopar @nmerket @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@Myoldmopar @nmerket @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@Myoldmopar @nmerket @lgentile it has been 14 days since this pull request was last updated.

Commit:
a few docs fixes and removing some debug printing

Commit:
Merge remote-tracking branch 'origin/develop' into empd

jdwoods21 committed:
If statements to catch prevent errors

Added if statements to MoistureBalanceEMPDManager.cc to limit mass flux
to surface and deep layers, and to limit RH of surface layer, based on
values of zone, surface, and deep layer RH levels from previous
timestep.

Commit:
Merge branch 'develop' into empd

Commit:
Adding autocalculate option for moisture penetration depths.

Commit:
removing changes to old transition rules spreadsheet

Commit:
updating ioref to reflect autocalc option on input

nmerket commented:
@Myoldmopar, this is ready to go again. 🎉  The only caveat is that the wonderful Fortran transition code you wrote for 8.4->8.5 is going to have to be moved into the 8.5->8.6 transition program. I think I took care of moving everything else (Rules*.csv), etc.

Myoldmopar committed:
Move transition program to 8.6 file, merge develop

Myoldmopar committed:
Removing executable bit from rules spreadsheet

Myoldmopar committed:
Force revert 84-85 rules file from develop

Myoldmopar committed:
Update style/format stuff

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into empd

Myoldmopar commented:
Hmm.  So @nmerket, what piece of this is causing the eso small diffs in pretty much every file?  It looks like it goes back to the early days of this PR, although at the beginning we weren't reporting as much detail on the CI results, so it wasn't obvious.  I'm just having trouble seeing which code change is triggering this in the non-empd files.  I expect the audit diffs and such, but the math diffs are unexpected unless you fixed something along the way.

nmerket commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -5289,7 +5289,7 @@ CalcHeatBalanceInsideSurf( Optional_int_const ZoneToResimulate ) // if passed in
 			Real64 const HConvIn_surf( HConvInFD( SurfNum ) = HConvIn( SurfNum ) );
 			RhoVaporAirIn( SurfNum ) = min( PsyRhovFnTdbWPb_fast( MAT_zone, ZoneAirHumRat_zone, OutBaroPress ), PsyRhovFnTdbRh( MAT_zone, 1.0, HBSurfManInsideSurf ) );
-			HMassConvInFD( SurfNum ) = HConvIn_surf / ( ( PsyRhoAirFnPbTdbW_fast( OutBaroPress, MAT_zone, ZoneAirHumRat_zone ) + RhoVaporAirIn( SurfNum ) ) * PsyCpAirFnWTdb_fast( ZoneAirHumRat_zone, MAT_zone ) );
+			HMassConvInFD( SurfNum ) = HConvIn_surf / ( PsyRhoAirFnPbTdbW_fast( OutBaroPress, MAT_zone, ZoneAirHumRat_zone ) * PsyCpAirFnWTdb_fast( ZoneAirHumRat_zone, MAT_zone ) );
```
@Myoldmopar, if I had to guess, I would say this is the culprit. Maybe I'll revert this one change and see if it gets rid of the ESO small diffs. 

@jdwoods21, Do you remember why we did this? Was it a bug fix?

jdwoods21 replied:
@Myoldmoparhttps://github.com/Myoldmopar, the equation below was changed because the previous equation was incorrect. The specific heat (PsyCpAir...) is defined on a dry-air basis [J/kg_da-K], so you want to multiply only by the density of air, (PsyRhoAir...) [kg_da/m^3]. Adding to this the vapor density (RhoVaporAirIn) puts the density on a total air/vapor mixture basis [kg_total/m^3], which is inconsistent with how the specific heat is defined, and thus gives the incorrect answer.

Commit:
reverting change to HeatBalanceSurfaceManager.cc to see if it removes the ESO diffs

jdwoods21 committed:
Updating EMPD section in engineering ref

jdwoods21 committed:
EMPD doc sections

Editing EMPD documentation and adding figure for EMPD documentation

Commit:
Revert "reverting change to HeatBalanceSurfaceManager.cc to see if it removes the ESO diffs"

This reverts commit c59d51ad042ca7eb84297210592c8b442fbffb9d.

Commit:
Merge remote-tracking branch 'origin/develop' into empd

Myoldmopar commented:
OK, so we need to decide on handling the bug fix.  My preference would be for you to re-apply the bug fix in an isolated branch that we can review and get merged in first, then the only diffs here would be what it shows right now, yet it would have the bug fix in place nicely.

jdwoods21 committed:
Update to EMPD docs

Hopefully the last one...

Commit:
Remove specific heat bugfix from this branch.
To be implemented on a separate branch. empd_bugfix.

This reverts commit ab43bbb15b29301d3161e02e4474919f3d6ea376.

Commit:
Merge branch 'develop' into empd

nmerket commented:
@Myoldmopar I removed the specific heat bug fix from this branch. This should be ready to go.
